File: songs/haskell/07-tags-unique.hs
exclude :: String -> ([String] -> [String])
exclude tag = filter((/=) tag)

unique :: [String] -> [String]
unique [] = []
unique (tag:tags) = tag:unique(exclude tag tags)

tags :: [String]
tags = ["rock", "jazz", "rock", "pop", "pop"]

main = print (unique tags)

File: songs/rust/07-tags-unique.rs
fn exclude<'lt>(
    value: &'lt str, tags: Vec<&'lt str>
) -> Vec<&'lt str> {
    let mut new_tags: Vec<&str> = Vec::new();

    for tag in tags.iter() {
        if tag != &value {
            new_tags.push(tag); 
        }
    }

    new_tags
}

fn unique<'lt>(
    tags: Vec<&'lt str>
) -> Vec<&'lt str> {
    if tags.len() <= 1 {
        tags
    } else {
        let head = &tags[0];
        let tail = tags[1..].to_vec();

        let mut new_tail = unique(exclude(head, tail));

        new_tail.insert(0, head);
        new_tail
    }
}

fn main() {
    let tags: Vec<&str> =
        vec!["rock", "jazz", "rock", "pop", "pop"];

    println!("{:?}", unique(tags));
}

